### PR TITLE
PS-5859: Merge MySQL 5.7.28 (part 1)

### PR DESCRIPTION
**Pull request by inikep**
https://jira.percona.com/browse/PS-5859
***
Updated man pages from MySQL Server 5.7.28 source tarball.
 ***
Updated 'scripts/fill_help_tables.sql' from MySQL Server 5.7.28 source tarball.
***
VERSION raised to "5.7.28-31".
univ.i version raised to "31".
***
Removed remaining code related to HAVE_YASSL
***
Reverted PS-5828 in favor of upstream solution.
`PS-5828: incorrect length in read_ok_ex() in sql-common/client.c:805`
https://jira.percona.com/browse/PS-5828 (commit e788bc55aa)
https://bugs.mysql.com/bug.php?id=96433 (commit mysql/mysql-server@c44b18041f)
***
Kept our solution for PS-5836 and reverted upstream.
`PS-5836: clang-8 reports broken source encoding in scripts/sql_commands_help_data.h`
https://jira.percona.com/browse/PS-5836 (commit e788bc55aa)
https://bugs.mysql.com/bug.php?id=96449 (commit mysql/mysql-server@19a667b6ac)
***
Reverted PS-3410 in favor of upstream solution with added our `innodb.percona_bulk_load_insert`.
`PS-3410: LP #1570114: Long running ALTER TABLE ADD INDEX causes semaphore wait > 600 assertion`
https://jira.percona.com/browse/PS-3410 (commit 7ebaef8e25)
https://bugs.mysql.com/bug.php?id=82940 (commit mysql/mysql-server@f9fb96c20f)  
***
Fixed gcc 9.2.1 warnings in `storage/innobase/log/log0online.cc` and `storage/innobase/os/os0file.cc`

### CODE REVIEW

**Comment by inikep** [comment]
Tested:
`-DWITH_ASAN=ON --big-test` at https://ps57.cd.percona.com/job/percona-server-5.7-asan-param/5/
`--big-test` at https://ps57.cd.percona.com/job/percona-server-5.7-param/180

File: storage/innobase/btr/btr0bulk.cc [hunk 1]
```diff
@@ -58,19 +58,11 @@ PageBulk::init()
 	mtr = static_cast<mtr_t*>(
 		mem_heap_alloc(m_heap, sizeof(mtr_t)));
 	mtr_start(mtr);
-<<<<<<< HEAD
```
Comment (percona-ysorokin) [inline storage/innobase/btr/btr0bulk.cc]: I believe https://github.com/percona/percona-server/commit/7ebaef8e255afe3ef88ee319871e2e51b322e2c7#diff-1b7266575f084a759d5bee343efe91d0R7022 should also be reverted

**Comment by percona-ysorokin** [comment]
@inikep in https://ps57.cd.percona.com/job/percona-server-5.7-asan-param/5/
MTR is run without `--sanitize` option
https://ps57.cd.percona.com/job/percona-server-5.7-pipeline/7491/consoleText
It suspiciously looks very clean :)

File: storage/innobase/btr/btr0bulk.cc (hunk 1 above)
Comment (inikep) [reply storage/innobase/btr/btr0bulk.cc]: It's also added by upstream at https://github.com/mysql/mysql-server/commit/f9fb96c20f#diff-1b7266575f084a759d5bee343efe91d0R6793

Comment (percona-ysorokin) [reply storage/innobase/btr/btr0bulk.cc]: Sorry, was looking a the wrong branch

**Comment by inikep** [comment]
@percona-ysorokin I followed our guide "Upstream merges" on Confluence to use `--big-test --sanitize (on 8.0)`.

**Comment by inikep** [comment]
`-DWITH_ASAN=ON --big-test --sanitize` at https://ps57.cd.percona.com/job/percona-server-5.7-asan-param/6
`-DWITH_VALGRIND=ON --big-test` at https://ps57.cd.percona.com/job/percona-server-5.7-valgrind-param/5

**Comment by percona-ysorokin** [comment]
> Upstream merges

Confluence document updated

**Comment by percona-ysorokin** [comment]
https://ps57.cd.percona.com/job/percona-server-5.7-pipeline/7525/consoleText
```
==26003==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 536 byte(s) in 1 object(s) allocated from:
    #0 0x7efffb305b50 in __interceptor_malloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdeb50)
    #1 0x5588b6614d8a in my_raw_malloc /tmp/ps/mysys/my_malloc.c:203
    #2 0x5588b6614d8a in my_malloc /tmp/ps/mysys/my_malloc.c:66
    #3 0x5588b73a3b44 in mysql_socket_vio_new /tmp/ps/vio/vio.c:299
    #4 0x5588b5ef6023 in Channel_info_local_socket::create_and_init_vio() const /tmp/ps/sql/conn_handler/socket_connection.cc:161
    #5 0x5588b6376ead in Channel_info::create_thd() /tmp/ps/sql/conn_handler/channel_info.cc:35
    #6 0x5588b5efb652 in Channel_info_local_socket::create_thd() /tmp/ps/sql/conn_handler/socket_connection.cc:176
    #7 0x5588b5ed411d in Thread_pool_connection_handler::add_connection(Channel_info*) /tmp/ps/sql/threadpool_unix.cc:1335
    #8 0x5588b609eefa in Connection_handler_manager::process_new_connection(Channel_info*) /tmp/ps/sql/conn_handler/connection_handler_manager.cc:322
    #9 0x5588b5f30dd9 in Connection_acceptor<Mysqld_socket_listener>::connection_event_loop() /tmp/ps/sql/conn_handler/connection_acceptor.h:75
    #10 0x5588b5f30dd9 in mysqld_main(int, char**) /tmp/ps/sql/mysqld.cc:5513
    #11 0x5588b4883772 in main /tmp/ps/sql/main.cc:32
    #12 0x7efff8b70b96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)
```
Also check if this is upstream

**Comment by percona-ysorokin** [comment]
https://ps57.cd.percona.com/job/percona-server-5.7-pipeline/7525/consoleText
```
==18788==ERROR: LeakSanitizer: detected memory leaks

Indirect leak of 2208 byte(s) in 1 object(s) allocated from:
    #0 0x7f42a4034b50 in __interceptor_malloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdeb50)
    #1 0x55a945fb4054 in ut_allocator<unsigned char>::allocate(unsigned long, unsigned char const*, char const*, bool, bool) /tmp/ps/storage/innobase/include/ut0new.h:367
    #2 0x55a945fb4054 in mem_heap_create_block_func(mem_block_info_t*, unsigned long, char const*, unsigned long, unsigned long) /tmp/ps/storage/innobase/mem/mem0mem.cc:310
    #3 0x55a945fb53e2 in mem_heap_add_block(mem_block_info_t*, unsigned long) /tmp/ps/storage/innobase/mem/mem0mem.cc:415
    #4 0x55a945fb58e5 in mem_heap_alloc /tmp/ps/storage/innobase/include/mem0mem.ic:212
    #5 0x55a945fb58e5 in mem_heap_dup(mem_block_info_t*, void const*, unsigned long) /tmp/ps/storage/innobase/mem/mem0mem.cc:67
    #6 0x55a945fb5a7f in mem_heap_strdup(mem_block_info_t*, char const*) /tmp/ps/storage/innobase/mem/mem0mem.cc:54
    #7 0x55a9465d7b82 in dict_mem_fill_index_struct /tmp/ps/storage/innobase/include/dict0mem.ic:59
    #8 0x55a9465d7b82 in dict_mem_index_create(char const*, char const*, unsigned long, unsigned long, unsigned long) /tmp/ps/storage/innobase/dict/dict0mem.cc:700
    #9 0x55a9461110b6 in row_merge_create_index(trx_t*, dict_table_t*, index_def_t const*, dict_add_v_col_t const*) /tmp/ps/storage/innobase/row/row0merge.cc:4290
    #10 0x55a945e8b6fb in prepare_inplace_alter_table_dict /tmp/ps/storage/innobase/handler/handler0alter.cc:4971
    #11 0x55a945e978da in ha_innobase::prepare_inplace_alter_table(TABLE*, Alter_inplace_info*) /tmp/ps/storage/innobase/handler/handler0alter.cc:6390
    #12 0x55a943e8c431 in handler::ha_prepare_inplace_alter_table(TABLE*, Alter_inplace_info*) /tmp/ps/sql/handler.cc:5147
    #13 0x55a94531e970 in mysql_inplace_alter_table /tmp/ps/sql/sql_table.cc:7845
    #14 0x55a945330f32 in mysql_alter_table(THD*, char const*, char const*, st_ha_create_information*, TABLE_LIST*, Alter_info*) /tmp/ps/sql/sql_table.cc:10324
    #15 0x55a94586a549 in Sql_cmd_alter_table::execute(THD*) /tmp/ps/sql/sql_alter.cc:355
    #16 0x55a945190a41 in mysql_execute_command(THD*, bool) /tmp/ps/sql/sql_parse.cc:5154
    #17 0x55a9451997f7 in mysql_parse(THD*, Parser_state*, bool) /tmp/ps/sql/sql_parse.cc:5927
    #18 0x55a94519d04a in dispatch_command(THD*, COM_DATA const*, enum_server_command) /tmp/ps/sql/sql_parse.cc:1539
    #19 0x55a9451a2d59 in do_command(THD*) /tmp/ps/sql/sql_parse.cc:1060
    #20 0x55a945494ea2 in handle_connection /tmp/ps/sql/conn_handler/connection_handler_per_thread.cc:325
    #21 0x55a945c1544b in pfs_spawn_thread /tmp/ps/storage/perfschema/pfs.cc:2198
    #22 0x7f42a39056da in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x76da)
```
Also check if this is upstream
Could be related to https://jira.percona.com/browse/PS-5992

**Comment by percona-ysorokin** [comment]
https://ps57.cd.percona.com/job/percona-server-5.7-pipeline/7525/consoleText
```
Direct leak of 10 byte(s) in 1 object(s) allocated from:
    #0 0x7fe1a32c5d38 in __interceptor_calloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xded38)
    #1 0x7fe18a10a641  (<unknown module>)
    #2 0x7fe18a10fbff  (<unknown module>)
    #3 0x7fe18a118630  (<unknown module>)
    #4 0x7fe18a0eee45  (<unknown module>)
    #5 0x7fe18a0f1c1e  (<unknown module>)
    #6 0x7fe18a0f9793  (<unknown module>)
    #7 0x7fe18a0ff19e  (<unknown module>)
    #8 0x7fe18a10b378  (<unknown module>)
    #9 0x7fe18a128ae0  (<unknown module>)
    #10 0x7fe18a0ee985  (<unknown module>)
    #11 0x7fe18a1c71c4  (<unknown module>)
    #12 0x7fe18a15d0fa  (<unknown module>)
    #13 0x7fe1a2b966da in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x76da)
```
At lease identify which MTR test case generates this.
Also check if this is upstream.

**Comment by percona-ysorokin** [comment]
https://ps57.cd.percona.com/job/percona-server-5.7-pipeline/7524/consoleText
```
SUMMARY: AddressSanitizer: stack-overflow (/usr/lib/x86_64-linux-gnu/libasan.so.4+0x5ecd8) 
Thread T25 created by T0 here:
    #0 0x7f47e6398d2f in __interceptor_pthread_create (/usr/lib/x86_64-linux-gnu/libasan.so.4+0x37d2f)
    #1 0x5647fe19e425 in pfs_spawn_thread_v1 /tmp/ps/storage/perfschema/pfs.cc:2249
    #2 0x5647fdafa39a in inline_mysql_thread_create /tmp/ps/include/mysql/psi/mysql_thread.h:1304
    #3 0x5647fdafa39a in Per_thread_connection_handler::add_connection(Channel_info*) /tmp/ps/sql/conn_handler/connection_handler_per_thread.cc:429
    #4 0x5647fdc90b09 in Connection_handler_manager::process_new_connection(Channel_info*) /tmp/ps/sql/conn_handler/connection_handler_manager.cc:322
    #5 0x5647fdb36839 in Connection_acceptor<Mysqld_socket_listener>::connection_event_loop() /tmp/ps/sql/conn_handler/connection_acceptor.h:75
    #6 0x5647fdb36839 in mysqld_main(int, char**) /tmp/ps/sql/mysqld.cc:5513
    #7 0x7f47e3caab96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)
```

**Comment by percona-ysorokin** [comment]
Updated Valgrind run (--big-test)
https://ps57.cd.percona.com/job/percona-server-5.7-valgrind-param/6/

**Comment by inikep** [comment]
Memory leaks are detected on `main.threadpool_debug`, `group_replication.gr_majority_loss_restart`, `innodb.alter_crash`.
Upstream bugs are `group_replication.gr_majority_loss_restart`, `innodb.alter_crash`.
The stack overflow is detected only for gcc-7.

**Comment by inikep** [comment]
Retested at https://ps57.cd.percona.com/job/percona-server-5.7-param/182/

**Review by percona-ysorokin — APPROVED** [review]
LGTM

**Pull request merged**